lib/node_modules/@stdlib/blas/base/cscal/test/test.ndarray.js

```diff
@@ -0,0 +1,323 @@
+var abs = require( '@stdlib/math/base/special/abs' );
+var cscal = require( './../lib/ndarray.js' );
+
+
+// TESTS //
+
+tape( 'main export is a function', function test( t ) {
+	t.ok( true, __filename );
+	t.strictEqual( typeof cscal, 'function', 'main export is a function' );
+	t.end();
+});
+
+tape( 'the function has an arity of 5', function test( t ) {
+	t.strictEqual( cscal.length, 5, 'arity of 5' );
+	t.end();
+});
+
+tape( 'the function scales elements from `x` by `cx`', function test( t ) {
+	var expected;
+	var delta;
+	var viewX;
+	var tol;
+	var ca;
+	var cx;
+	var k;
+
+	cx = new Complex64Array( [
+		0.3, // 1
+		0.1, // 1
+		0.5, // 2
+		0.0, // 2
+		0.0, // 3
+		0.5, // 3
+		0.0, // 4
+		0.2, // 4
+		2.0,
+		3.0,
+		2.0,
+		3.0
+	] );
+	ca = new Complex64( 0.4, -0.7 );
+
+	cscal( 4, ca, cx, 1, 0 );
+
+	viewX = new Float32Array( cx.buffer );
+	expected = new Float32Array( [
+		0.19,  // 1
+		-0.17, // 1
+		0.2,   // 2
+		-0.35, // 2
+		0.35,  // 3
+		0.2,   // 3
+		0.14,  // 4
+		0.08,  // 4
+		2.0,
+		3.0,
+		2.0,
+		3.0
+	] );
+	for ( k = 0; k < expected.length; k++ ) {
+		if ( viewX[ k ] === expected[ k ] ) {
+			t.strictEqual( viewX[ k ], expected[ k ], 'returns expected value' );
+		}
+		else {
+			delta = abs( viewX[ k ] - expected[ k ] );
+			tol = 1.0 * EPS * abs( expected[ k ] );
+			t.ok( delta <= tol, 'within tolerance. x: '+viewX[ k ]+'. expected: '+expected[ k ]+'. delta: '+delta+'. tol: '+tol+'.' );
+		}
+	}
+	t.end();
+});
+
+tape( 'the function supports an `x` stride', function test( t ) {
+	var expected;
+	var delta;
+	var viewX;
+	var tol;
+	var ca;
+	var cx;
+	var k;
+
+	cx = new Complex64Array( [
+		0.1,  // 1
+		0.1,  // 1
+		3.0,
+		6.0,
+		-0.6, // 2
+		0.1,  // 2
+		4.0,
+		7.0,
+		0.1,  // 3
+		-0.3, // 3
+		7.0,
+		2.0
+	] );
+	ca = new Complex64( 0.4, -0.7 );
+
+	cscal( 3, ca, cx, 2, 0 );
+
+	viewX = new Float32Array( cx.buffer );
+	expected = new Float32Array( [
+		0.11,  // 1
+		-0.03, // 1
+		3.0,
+		6.0,
+		-0.17, // 2
+		0.46,  // 2
+		4.0,
+		7.0,
+		-0.17, // 3
+		-0.19, // 3
+		7.0,
+		2.0
+	] );
+	for ( k = 0; k < expected.length; k++ ) {
+		if ( viewX[ k ] === expected[ k ] ) {
+			t.strictEqual( viewX[ k ], expected[ k ], 'returns expected value' );
+		}
+		else {
+			delta = abs( viewX[ k ] - expected[ k ] );
+			tol = 1.0 * EPS * abs( expected[ k ] );
+			t.ok( delta <= tol, 'within tolerance. x: '+viewX[ k ]+'. expected: '+expected[ k ]+'. delta: '+delta+'. tol: '+tol+'.' );
+		}
+	}
+	t.end();
+});
+
+tape( 'the function supports an `x` offset', function test( t ) {
+	var expected;
+	var delta;
+	var viewX;
+	var tol;
+	var ca;
+	var cx;
+	var k;
+
+	cx = new Complex64Array( [
+		0.1,
+		0.1,
+		3.0,
+		6.0,
+		-0.6,
+		0.1,
+		4.0,  // 1
+		6.0,  // 1
+		0.1,  // 2
+		-0.3, // 2
+		7.0,  // 3
+		2.0   // 3
+	] );
+	ca = new Complex64( 0.4, -0.7 );
+
+	cscal( 3, ca, cx, 1, 3 );
+
+	viewX = new Float32Array( cx.buffer );
+	expected = new Float32Array( [
+		0.1,
```

Comment:
Question: are these values pulled from the BLAS test suite? Or did you make these up? It's fine if the latter. But, if so, you can make your testing life easier by making integer values and thus not require approximate equality tests. If the former, then, yes, approx equality tests are necesary.

Reply:
Since there are no tests for the offset case in the BLAS test suite, I took the input data ca and cx from the test suite and performed the operation myself with an offset.  

Reply:
That makes sense and justifies the use of approximate equality tests. Thanks for following up.